### PR TITLE
Expand documentation

### PR DESCRIPTION
**Pull request by Gastove**
*Note*: to get everything working nicely, this branch was opened off of my branch for #97. It will be much easier to review this PR after #97 has closed, as it contains all of 97s commits!

This PR proposes some expansions to Fleece's documentation. In particular, my goal is to cover sets of use-cases I found non-obvious as a newcomer to Fleece. I've tried to clarify a few points around working with DUs, as well as dealing with fields that are required in a given JSON object but _not_ in F#. 

Per #98, I've also bumped the SDK version in the project `global.json` to 5.0, to allow the use of F#5's package-loading features in documentation scripts.

I am not, for the record, adamant about any of this. Fleece is a great project, and I'd love to help! And, I'm pretty happy to take a different approach.

### CODE REVIEW

**Comment by Gastove** [comment]
Huh. I just realized I never bumped the actual dotnet versions being used for CI. But -- CI is perfectly happy, so it certainly doesn't seem to be harming anything. If the `global.json` change works for y'all, I'm very happy to make sure CI stays in sync as part of this PR. Just let me know!

**Comment by wallymathieu** [comment]
I think it makes sense to upgrade in order to be able to use this new feature 👍. I think you can also remove the:

- docsrc/tool/download_nugets.ps1
- docsrc/tool/download_nugets.sh

since these download needed nuget packages into folders, while if you use the new F# 5 nuget reference syntax these are unnecessary.

File: docsrc/content/codec.fsx [hunk 1]
```diff
@@ -1,11 +1,10 @@
 (*** hide ***)
 // This block of code is omitted in the generated HTML documentation. Use
 // it to define helpers that you do not want to show in the documentation.
-#r @"../../src/Fleece.SystemJson/bin/Release/netstandard2.1/System.Json.dll"
```
Comment (wallymathieu) [inline docsrc/content/codec.fsx:4]: The code you replace is actually the compiled binary, while here you change it to be the released library version

**Comment by wallymathieu** [comment]
Note for instance:

https://github.com/fsprojects/Fleece/blob/master/docsrc/tool/generate.fs#L52-L73

```f#
let docPackagePath  path =
    __SOURCE_DIRECTORY__ + @"/../../packages/docs/" + path
...
let evaluationOptions = 
    [| 
         includeDir "FSharp.Core/lib/netstandard2.0/"
         reference "FSharp.Core/lib/netstandard2.0/FSharp.Core.dll" 
         includeDir "FSharp.Literate/lib/netstandard2.0/" 
         includeDir "MathNet.Numerics/lib/netstandard2.0/" 
         reference "MathNet.Numerics/lib/netstandard2.0/MathNet.Numerics.dll" 
         includeDir "MathNet.Numerics.FSharp/lib/netstandard2.0/" 
         reference "MathNet.Numerics.FSharp/lib/netstandard2.0/MathNet.Numerics.FSharp.dll" 
         includeDir "System.Buffers/lib/netstandard2.0/" 
         includeDir "System.Collections.Immutable/lib/netstandard2.0/" 
         includeDir "System.Reflection.Metadata/lib/netstandard2.0/" 
         includeDir "System.ValueTuple/lib/netstandard1.0/" 
         reference "System.ValueTuple/lib/netstandard1.0/System.ValueTuple.dll" 
         includeDir "FSharp.Compiler.Service/lib/netstandard2.0/" 
         reference "FSharp.Compiler.Service/lib/netstandard2.0/FSharp.Compiler.Service.dll" |] 
```

you could use F#5 NuGet reference instead of using NuGet to download packages. I think that it would then only be necessary to reference `Suave`, `TaskBuilder.fs` et.c. directly in the documentation.

**Comment by wallymathieu** [comment]
In order to test out such changes you could do:
 
```bash
dotnet tool restore
dotnet build -c Release
dotnet run --project ./docsrc/tool
```

**Comment by wallymathieu** [comment]
Though perhaps I (or someone willing) should do a separate pull request (once you are finished with this one) with those changes since I think the scope you have for this is to expand the documentation.

**Comment by wallymathieu** [comment]
Since the PR #102 has been merged, is there anything you want to change in this PR?

**Comment by Gastove** [comment]
Ha, I lost track of this. Will review, fix conflicts, etc.

**Comment by Gastove** [comment]
Okay, that should at least be properly rebased and cleaned up. I had a rather alarming editor snafu, so that was a bit messier than expected, but I think it looks OK now. 

I think the last thing I'd love y'all's guidance on is: using released library code versus compiled local assemblies in the docs. Using released library code is considerably more ergonomic; a human contributor doesn't have to remember to _do_ anything before the docs load correctly in an editor. On the other hand, this means there isn't actually exact parity between the docs and the code they document, which doesn't seem great. 

A further challenge: when referencing local assemblies, if one updates a dependency version for a particular library, all documentation paths must be updated to the same version. This does enforce correctness, and is just frustrating. 

My current thought is: 
- Nuget reference syntax with versions for external dependencies. (Makes it, if nothing else, easier to spot and update the versions.)
- Local DLL reference for Fleece itself. 
- Documentation making it clear to run a build before loading doc fsx files. 

Thoughts from y'all maintainers? What do you prefer?

**Comment by Gastove** [comment]
(Whatever we decide, I'll write some documentation about how to write/contribute documentation and add it to this PR, make sure it's captured clearly.)

**Comment by wallymathieu** [comment]
I think it makes sense to compile the documentation against the source, while having F# 5 NuGet references visible in the docs (for instance via markdown code includes in comments). Though having the docs compile against a released version would imply a separate life cycle of the docs compared compared to the lib: Since you'd have to document new features after a finished release. The reason to tie docs with locally compiled binaries is that you could start document unreleased features.

**Comment by wallymathieu** [comment]
I'm thinking to adjust this PR slightly in order to merge and push documentation. I'll see when I have the time.

**Review by wallymathieu — APPROVED** [review]
👍

**Comment by wallymathieu** [comment]
I've changed the ShapeD codec definition slightly, what's your thoughts @Gastove ?

File: docsrc/content/combinators.fsx [hunk 2]
```diff
@@ -1,13 +1,18 @@
 (*** hide ***)
 // This block of code is omitted in the generated HTML documentation. Use
 // it to define helpers that you do not want to show in the documentation.
-#r "nuget: FSharpPlus"
-#r "nuget: System.Json"
+#r "nuget: System.Json, 4.7.1"
+#r "nuget: FSharpPlus, 1.1.1"
```
Comment (gusty) [inline docsrc/content/combinators.fsx:5]: Shouldn't be 1.2.1 ?

File: docsrc/content/comparison-with-json-net.fsx [hunk 3]
```diff
@@ -1,11 +1,16 @@
 (*** hide ***)
 // This block of code is omitted in the generated HTML documentation. Use
 // it to define helpers that you do not want to show in the documentation.
-#r @"nuget: Newtonsoft.Json"
-#r "nuget: FSharpPlus, 1.1.7"
-#r @"../../src/Fleece.SystemJson/bin/Release/netstandard2.1/Fleece.SystemJson.dll"
+#r "nuget: Newtonsoft.Json, 10.0.2"
+#r "nuget: FSharpPlus, 1.1.1"
```
Comment (gusty) [inline docsrc/content/comparison-with-json-net.fsx:5]: Same here

**Comment by wallymathieu** [comment]
I'm tinking of fixing the issues in order to be able to merge this PR @Gastove

File: docsrc/content/codec.fsx [hunk 4]
```diff
@@ -1,18 +1,24 @@
 (*** hide ***)
 // This block of code is omitted in the generated HTML documentation. Use
 // it to define helpers that you do not want to show in the documentation.
-#r "nuget: FSharpPlus"
-#r "nuget: System.Json"
+#r "nuget: System.Json, 4.7.1"
+#r "nuget: FSharpPlus, 1.2.1"
```
Comment (gusty) [inline docsrc/content/codec.fsx:5]: Let's use 1.2.2 now

File: docsrc/content/combinators.fsx [hunk 5]
```diff
@@ -1,13 +1,18 @@
 (*** hide ***)
 // This block of code is omitted in the generated HTML documentation. Use
 // it to define helpers that you do not want to show in the documentation.
-#r "nuget: FSharpPlus"
-#r "nuget: System.Json"
+#r "nuget: System.Json, 4.7.1"
+#r "nuget: FSharpPlus, 1.2.1"
```
Comment (gusty) [inline docsrc/content/combinators.fsx:5]: Same here

File: docsrc/content/comparison-with-json-net.fsx [hunk 6]
```diff
@@ -1,11 +1,16 @@
 (*** hide ***)
 // This block of code is omitted in the generated HTML documentation. Use
 // it to define helpers that you do not want to show in the documentation.
-#r @"nuget: Newtonsoft.Json"
-#r "nuget: FSharpPlus, 1.1.7"
-#r @"../../src/Fleece.SystemJson/bin/Release/netstandard2.1/Fleece.SystemJson.dll"
+#r "nuget: Newtonsoft.Json, 10.0.2"
+#r "nuget: FSharpPlus, 1.2.1"
```
Comment (gusty) [inline docsrc/content/comparison-with-json-net.fsx:5]: .. and here

**Review by gusty — APPROVED** [review]
Approved with suggestions: we have to deprecate some stuff, but we can do it later, in the PR for Fleece vNext.

File: docsrc/content/codec.fsx (hunk 4 above)
Comment (wallymathieu) [reply docsrc/content/codec.fsx:5]: Sounds good

File: docsrc/content/further-techniques.fsx [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+type Response =
+    { Result: string option
+      Error: string option }
+      static member JsonObjCodec =
+          fun r e _ -> { Result = r; Error = e }
+          |> withFields
+          |> jfieldOpt "result"  (fun r -> r.Result)
+          |> jfieldOpt "error"   (fun r -> r.Error)
+          |> jfield    "jsonrpc" (fun _ -> "2.0")
```
[the remark this answers is not in the document]
Comment (wallymathieu) [reply docsrc/content/further-techniques.fsx:58]: Sounds good. Should we include it in the docs and write that it's obsolete then?

**Comment by wallymathieu** [comment]
Many thanks @Gastove for the contribution!

**Pull request merged**